common/scheme_plugin.go

```diff
@@ -148,32 +206,76 @@ func (s *SchemeRPC) GetFormat() AttestationFormat {
 	var resp AttestationFormat
 	err := s.client.Call("Plugin.GetFormat", new(interface{}), &resp)
 	if err != nil {
-		log.Printf("ERROR during GetName RPC: %v\n", err)
+		log.Printf("ERROR during GetFormat RPC: %v\n", err)
 		return AttestationFormat_UNKNOWN_FORMAT
 	}
 	return resp
 }
 
-func (s *SchemeRPC) GetTrustAnchorID(token *AttestationToken) string {
+func (s *SchemeRPC) SynthKeysFromSwComponent(tenantID string, swComp *SwComponent) ([]string, error) {
+	var (
+		err  error
+		resp []string
+		args SynthKeysArgs
+	)
+
+	args.TenantID = tenantID
+
+	args.EndorsementJSON, err = json.Marshal(swComp)
+	if err != nil {
+		return nil, fmt.Errorf("failed unmarshaling response from RPC server: %w", err)
+	}
+
+	err = s.client.Call("Plugin.SynthKeysFromSwComponent", args, &resp)
+	if err != nil {
+		return nil, fmt.Errorf("failed call to SynthKeysFromSwComponent: %w", err)
+	}
+
+	return resp, nil
+}
+
+func (s *SchemeRPC) SynthKeysFromTrustAnchor(tenantID string, ta *TrustAnchor) ([]string, error) {
```

Comment:
> Shouldn't this be returning the same thing as `GetTrustAnchorID` below, with the only difference being the input it operates on?

Not really.  `GetTrustAnchorID` uses one specific name of the object it needs to retrieve, whereas the `SynthKeysFrom...` may potentially create different aliases for the same underlying object to be used in different queries, i.e., not just `GetTrustAnchorID` but also the other provisioning APIs which may have different access patterns.
 
> Also, for the sake of consistency in the APIs, we should also align on whether the `tenantID` is assumed to be part of a structure (as with `GetTrustAnchorID` below), or is always passed as the first argument.

True.

> (Note: I don't have a strong preference either way, and am happy for the tenant ID to be removed from the `AttestationToken` ).

I don't have either :-)

Reply:
> Not really. GetTrustAnchorID uses one specific name of the object it needs to retrieve, whereas the SynthKeysFrom... may potentially create different aliases for the same underlying object to be used in different queries, i.e., not just GetTrustAnchorID but also the other provisioning APIs which may have different access patterns.

Mm OK, I think I have a different intuition for what is implied by "key". I gather this intended to be used more like indexes or query parameters? Also, if they are intended to be used for different quires and in different contexts, how does one identify which query/context a particular result is to be used in? Is this intended to be implicit in the positional index of the value within the slice? If so, perhaps it would be clearer to make it explicit by making the different queries fields in a struct, or (if this is going to be scheme-specific) keys in a map?

Reply:
> [...] or (if this is going to be scheme-specific) keys in a map?

That's a really intriguing idea, I mean this notion of storing the key purpose (i.e., the associated query) alongside the key itself.

The reason I made it an dumb array - for now - was for minimising the "need to know."  In my view it's only the scheme-specific logics (i.e., the plugin) that knows which keys it'll need to use, and therefore has to synthesise.  Whilst what the "core" VTS part that populates the kvstore needs to know is only that the supplied "val" will be aliased under the associated 1..n different "key"s.

Reply:
Sure, and with a map you're still have the need-to-know as only the scheme knows how to interpret they keys, and from the out side code, it's basically an opaque `interface{}` either way. The textual, rather than numerical indexes just make make it easier for debugging.

Anyway, no biggie; just a thought. An array is totally fine for now.